src/rust/ide/src/controller/searcher.rs

```diff
@@ -0,0 +1,265 @@
+//! This module contains all structures related to Searcher Controller.
+
+use crate::prelude::*;
+
+use crate::notification;
+
+use data::text::TextLocation;
+use enso_protocol::language_server;
+use flo_stream::Subscriber;
+
+
+
+// =======================
+// === Suggestion List ===
+// =======================
+
+/// A single suggestion on the Searcher suggestion list.
+#[derive(Clone,CloneRef,Debug,Eq,PartialEq)]
+pub enum Suggestion {
+    /// Suggestion for input completion: possible functions, arguments, etc.
+    Completion(Rc<model::suggestion_database::Entry>)
+    // In future, other suggestion types will be added (like suggestions of actions, etc.).
+}
+
+/// List of suggestions available in Searcher.
+#[derive(Clone,CloneRef,Debug)]
+pub enum Suggestions {
+    /// The suggestion list is still loading from the Language Server.
+    Loading,
+    /// The suggestion list is loaded.
+    #[allow(missing_docs)]
+    Loaded {
+        list : Rc<Vec<Suggestion>>
+    },
+    /// Loading suggestion list results in error.
+    Error(Rc<failure::Error>)
+}
+
+impl Suggestions {
+    /// Check if suggestion list is still loading.
+    pub fn is_loading(&self) -> bool {
+        match self {
+            Self::Loading => true,
+            _             => false,
+        }
+    }
+
+    /// Check if retrieving suggestion list was unsuccessful
+    pub fn is_error(&self) -> bool {
+        match self {
+            Self::Error(_) => true,
+            _              => false,
+        }
+    }
+
+    /// Get the list of suggestions. Returns None if still loading or error was returned.
+    pub fn list(&self) -> Option<&Vec<Suggestion>> {
+        match self {
+            Self::Loaded {list} => Some(list),
+            _                   => None,
+        }
+    }
+}
+
+impl Default for Suggestions {
+    fn default() -> Self {
+        Self::Loading
+    }
+}
+
+
+
+// =====================
+// === Notifications ===
+// =====================
+
+/// The notification emitted by Searcher Controller
+#[derive(Copy,Clone,Debug,Eq,PartialEq)]
+pub enum Notification {
+    /// A new Suggestion list is available.
+    NewList
```

Comment:
Notification::NewList does not tell anything about the factthat this is a __suggestion__ list